### ISSUE TITLE
DMatrix from Numpy/Pandas vs CSV file produces different prediction

### ISSUE THREAD

**Issue by aws-patlin** [dmlc/xgboost#5456]
I have found that for a model trained on data with negative values, predicting on a DMatrix created from data containing NaNs loaded first into a numpy array or pandas dataframe produces a different prediction compared to creating a DMatrix directly from the CSV file. I've observed this behavior in the Python API for both XGBoost 0.90 and 1.0.0 (haven't tested older versions). Here is a minimal script to reproduce:

```
import numpy as np
import pandas as pd
import sklearn.metrics as sm
import xgboost as xgb


# Create data
n = np.random.uniform(low=-100, high=1000, size=[250,115])  # Contains negative values
l = np.random.randint(0,high=2,size=250)
df = pd.DataFrame(n)
df[0] = l
df.fillna('').to_csv('train.csv', header=False, index=False)

n = np.random.uniform(low=0, high=1000, size=[100,115])
# n[1::2,1:11] = np.nan  # Add NaNs, uncomment to produce inaccurate predictions from Numpy/Pandas
l = np.random.randint(0,high=2,size=100)
df = pd.DataFrame(n)
df[0] = l
df.fillna('').to_csv('val.csv', header=False, index=False)


# Train booster
dtrain = xgb.DMatrix("train.csv?format=csv&label_column=0&delimiter=,")
dval = xgb.DMatrix("val.csv?format=csv&label_column=0&delimiter=,")
params = {
    "max_depth": 8,
    "learning_rate": 0.03266812195144707,
    "objective": 'binary:logistic',
    "verbosity": 1,
    "reg_alpha": 0.7812482958179122,
    "reg_lambda": 1.8180791719825637,
    "colsample_bytree": 0.8,
    "subsample": 0.75,
    "min_child_weight": 0,
    "eval_metric": "auc"
}
watchlist = [(dtrain, 'train'), (dval, 'validation')]
booster = xgb.train(params, dtrain, num_boost_round=10, evals=watchlist)


# Predict on data read into Python data structures
val = pd.read_csv('val.csv', header=None)
x_val, y_val = val.iloc[:, 1:], val.iloc[:, 0]

pd_dm = xgb.DMatrix(x_val)
pd_dm.feature_names = ['f{}'.format(int(x)-1) for x in pd_dm.feature_names]  # Fix feature names
np_dm = xgb.DMatrix(x_val.to_numpy())

pd_pred = booster.predict(pd_dm)
np_pred = booster.predict(np_dm)

print("DMatrix from Pandas dataframe: ", sm.roc_auc_score(y_val, pd_pred))  # Incorrect AUC scores
print("DMatrix from Numpy array: ", sm.roc_auc_score(y_val, np_pred))
np.savetxt('np_pred', np_pred)


# Predict on data read from file
y_pred = booster.predict(dval)
print("DMatrix from original file: ", sm.roc_auc_score(y_val, y_pred))  # Correct AUC score
np.savetxt('y_pred', y_pred)


# Execute `diff y_pred np_pred` in the CLI to see that the predictions differ on the rows with NaNs
```

Example output:
```
[01:42:04] 250x114 matrix with 28500 entries loaded from train.csv?format=csv&label_column=0&delimiter=,
[01:42:04] 100x114 matrix with 11400 entries loaded from val.csv?format=csv&label_column=0&delimiter=,
[0]	train-auc:0.89699	validation-auc:0.50913
[1]	train-auc:0.95545	validation-auc:0.45982
[2]	train-auc:0.97545	validation-auc:0.50893
[3]	train-auc:0.98784	validation-auc:0.45252
[4]	train-auc:0.99488	validation-auc:0.44379
[5]	train-auc:0.99642	validation-auc:0.46956
[6]	train-auc:0.99834	validation-auc:0.45860
[7]	train-auc:0.99917	validation-auc:0.47119
[8]	train-auc:0.99962	validation-auc:0.43994
[9]	train-auc:0.99994	validation-auc:0.43222
DMatrix from Pandas dataframe:  0.4553571428571428
DMatrix from Numpy array:  0.4553571428571428
DMatrix from original file:  0.432224025974026
```

I'm not sure if this is a bug in the XGBoost Python API, or if this is caused by some issue in Python itself, since creating a DMatrix from a file basically just calls the C++ API.

**Comment by aws-patlin** [comment]
Was this fix confirmed to resolve the provided example?

For my understanding, was the issue that zeros were being pushed into the DMatrix for validation/prediction when reading from file, and this was producing an inaccurate prediction? Do we have an understanding of why this only seems to be a problem when negative values are present in the training data?

**Comment by hcho3** [comment]
@aws-patlin Yes, previously the CSV parser was pushing zero for all missing values. So `3,,,4` would be parsed into `[3, 0, 0, 4]`, not `[3, nan, nan, 4]` (here I use `nan` to represent missing value).

**Comment by trivialfis** [comment]
Sorry, seems being closed by github automatically.  We still need #5466 for the complete fix.  The issue is in dmlc CSV parser.  It adds a 0 when entry is missing:
```
1,,,,3
```
 results in a row with value: 
```
1, 0, 0, 0, 0, 3
```

**Comment by aws-patlin** [comment]
Here is what I think is going on (please correct me if I'm wrong, since my understanding of XGBoost/decision trees is pretty rudimentary):

When negative values are not present in the training set, filling in zeros for NaNs may not produce any visible difference if zeros and NaNs always fall on the same side of the trees' splits. For example, if a split is f1 < 10, and both 0 and NaN count as True.

But when negative values are present, this may no longer be the case. For example, if a split is now f1 < -5, and NaN counts as True, but 0 counts as False.

**Comment by aws-patlin** [comment]
If this was automatically closed before the complete fix has been released, can @hcho3 re-open it? Also, will this bug fix go into older releases like 0.90, or is it only going into the latest release?

**Comment by hcho3** [comment]
We do not backport bug fixes to old releases in most circumstances.
